flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java

```diff
@@ -0,0 +1,177 @@
+        this.sizeLimit = sizeLimit;
+        this.total = 0L;
+        this.caches = new LinkedHashMap<>(INITIAL_SIZE, 0.75F, true);
+        this.blobKeyByJob = new HashMap<>();
+    }
+
+    /**
+     * Check the size limit and return the files to delete.
+     *
+     * @param size size of the target file intended to put into cache
+     * @return list of files to delete before saving the target file
+     */
+    public List<Tuple2<JobID, BlobKey>> checkLimit(long size) {
+        checkArgument(size >= 0);
+
+        synchronized (lock) {
+            List<Tuple2<JobID, BlobKey>> blobsToDelete = new ArrayList<>();
+
+            long current = total;
+
+            for (Map.Entry<Tuple2<JobID, BlobKey>, Long> entry : caches.entrySet()) {
+                if (current + size > sizeLimit) {
+                    blobsToDelete.add(entry.getKey());
+                    current -= entry.getValue();
+                }
+            }
+
+            return blobsToDelete;
+        }
+    }
+
+    /** Register the target file to the tracker. */
+    public void track(@Nullable JobID jobId, BlobKey blobKey, long size) {
+        checkNotNull(blobKey);
+        checkArgument(size >= 0);
+
+        synchronized (lock) {
+            caches.put(Tuple2.of(jobId, blobKey), size);
+            if (jobId != null) {
+                blobKeyByJob.computeIfAbsent(jobId, ignore -> new HashSet<>()).add(blobKey);
+            }
+            total += size;
+            if (total > sizeLimit) {
+                LOG.warn(
+                        "The overall size of ShuffleDescriptors in PermanentBlobCache exceeds the limit. "
+                                + "Limit: {}, current: {}. The size of next ShuffleDescriptors: {}.",
+                        sizeLimit,
+                        total,
+                        size);
+            }
+        }
+    }
+
+    /** Remove the BLOB from the tracker. */
+    public void untrack(Tuple2<JobID, BlobKey> key) {
+        checkNotNull(key);
+
+        synchronized (lock) {
+            if (key.f0 != null) {
+                blobKeyByJob.computeIfAbsent(key.f0, ignore -> new HashSet<>()).remove(key.f1);
+            }
+            Long size = caches.remove(key);
+            if (size != null && size >= 0) {
+                total -= size;
+            }
+        }
+    }
+
+    /** Remove the BLOB from the tracker. */
+    public void untrack(@Nullable JobID jobId, BlobKey blobKey) {
+        checkNotNull(blobKey);
+
+        untrack(Tuple2.of(jobId, blobKey));
+    }
+
+    /**
+     * Update the last used index for the BLOBs so that the tracker can easily find out the last
+     * recently used BLOBs.
+     */
+    public void update(@Nullable JobID jobId, BlobKey blobKey) {
+        checkNotNull(blobKey);
+
+        synchronized (lock) {
+            caches.get(Tuple2.of(jobId, blobKey));
+        }
+    }
+
+    /** Unregister all the tracked BLOBs related to given job. */
+    public void unregisterJob(JobID jobId) {
```

Comment:
I prefer the name to be `untrackAll` for easier understanding and also because there is no `registerJob()`